File: chat-client/src/agent.ts
import { ChatOpenAI } from "@langchain/openai";
import { ChatAnthropic } from "@langchain/anthropic";
import { createReactAgent } from "@langchain/langgraph/prebuilt";
import { HumanMessage, AIMessage, AIMessageChunk, ToolMessage, BaseMessage } from "@langchain/core/messages";
import type { StructuredToolInterface } from "@langchain/core/tools";
import type { BaseChatModel } from "@langchain/core/language_models/chat_models";


const SYSTEM_PROMPT = `You are an engaging Dungeon Master running a text-based RPG adventure.

## Communication Modes

Players communicate in TWO modes, indicated by message prefixes:

**[OOC]** (Out-of-Character) - The player is talking to YOU as the game master:
- Answer questions about game mechanics, available tools, rules
- Explain character stats, inventory, abilities
- Discuss strategy, give hints if asked
- Help with session management (creating sessions, checking state)
- Be helpful and informative, break the fourth wall freely
- List available actions or explain how things work

**[IC]** (In-Character) - The player is roleplaying as their character:
- Respond as the world and NPCs would
- Describe scenes, environments, and NPC reactions dramatically
- Stay fully immersed - never break character
- Narrate the results of their actions cinematically
- Roll dice when needed and describe outcomes vividly

## Your Role

- Create immersive descriptions of locations, NPCs, and events
- Guide players through the story using the available tools
- Roll dice for combat, skill checks, and random outcomes
- Keep track of the game state using sessions and character sheets
- Reference world lore for consistent storytelling

## Getting Started with a New Player

1. Create a session with create_session() if one doesn't exist
2. Show them the current map with get_current_map()
3. Describe their surroundings vividly
4. Ask what they'd like to do

When in IC mode, be descriptive, dramatic, and fun!
Use the lookup tools to reference lore and keep the world consistent.
When combat happens, use roll_dice and describe the action cinematically.`;

export type ModelProvider = "openai" | "anthropic" | "openrouter";

export interface ModelConfig {
  provider: ModelProvider;
  model?: string;
}

export interface DebugCallbacks {
  onThinking?: (text: string) => void;
  onToolCall?: (name: string, input: Record<string, unknown>) => void;
  onToolResult?: (name: string, result: string) => void;
}

export interface GameAgent {
  chat: (message: string) => Promise<string>;
  getHistory: () => BaseMessage[];
}

// Debug logging helper
const DEBUG = process.env.DEBUG_AGENT === "true";
function debugLog(stage: string, ...args: unknown[]) {
  if (DEBUG) {
    console.log(`[DEBUG ${new Date().toISOString()}] ${stage}:`, ...args);
  }
}

function createModel(config: ModelConfig): BaseChatModel {
  debugLog("createModel", "Creating model with config:", config);
  
  switch (config.provider) {
    case "anthropic":
      debugLog("createModel", "Using Anthropic provider");
      return new ChatAnthropic({
        model: config.model || "claude-sonnet-4-20250514",
        temperature: 0.7,
      });
    case "openrouter":
      debugLog("createModel", "Using OpenRouter provider with baseURL:", "https://openrouter.ai/api/v1");
      debugLog("createModel", "API key present:", !!process.env.OPENROUTER_API_KEY);
      debugLog("createModel", "Model:", config.model || "xiaomi/mimo-v2-flash:free");
      return new ChatOpenAI({
        model: config.model || "xiaomi/mimo-v2-flash:free",
        temperature: 0.7,
        streaming: true,
        apiKey: process.env.OPENROUTER_API_KEY,
        configuration: {
          baseURL: "https://openrouter.ai/api/v1",
        },
      });
    case "openai":
    default:
      debugLog("createModel", "Using OpenAI provider");
      return new ChatOpenAI({
        model: config.model || "gpt-4o",
        temperature: 0.7,
      });
  }
}

export async function createGameAgent(
  tools: StructuredToolInterface[],
  config: ModelConfig = { provider: "openrouter" },
  debugCallbacks?: DebugCallbacks
): Promise<GameAgent> {
  debugLog("createGameAgent", "Starting agent creation with provider:", config.provider);
  debugLog("createGameAgent", "Number of tools:", tools.length);
  
  const model = createModel(config);
  debugLog("createGameAgent", "Model created successfully");

  debugLog("createGameAgent", "Creating ReAct agent...");
  const agent = createReactAgent({
    llm: model,
    tools,
    messageModifier: SYSTEM_PROMPT,
  });
  debugLog("createGameAgent", "ReAct agent created successfully");

  let messageHistory: BaseMessage[] = [];

  return {
    chat: async (message: string): Promise<string> => {
      debugLog("chat", "Starting chat with message:", message.slice(0, 100) + "...");
      messageHistory.push(new HumanMessage(message));

      debugLog("chat", "Calling agent.stream()...");
      // Use streaming to capture intermediate steps
      let stream;
      try {
        stream = await agent.stream(
          { messages: messageHistory },
          { streamMode: "updates" }
        );
      } catch (streamError: unknown) {
        debugLog("chat", "ERROR in agent.stream():", streamError);
        // Try to extract more details from the error
        const err = streamError as Error & { 
          response?: { status?: number; statusText?: string; data?: unknown };
          cause?: unknown;
          code?: string;
        };
        if (DEBUG) {
          console.error("[STREAM ERROR] Message:", err.message);
          console.error("[STREAM ERROR] Code:", err.code);
          console.error("[STREAM ERROR] Cause:", err.cause);
          if (err.response) {
            console.error("[STREAM ERROR] Response status:", err.response.status, err.response.statusText);
            console.error("[STREAM ERROR] Response data:", JSON.stringify(err.response.data, null, 2));
          }
          console.error("[STREAM ERROR] Full error:", err);
        }
        throw streamError;
      }
      debugLog("chat", "agent.stream() returned, starting iteration...");

      let finalResponse = "";
      const newMessages: BaseMessage[] = [];
      let chunkCount = 0;

      try {
      for await (const chunk of stream) {
        chunkCount++;
        debugLog("stream", `Received chunk #${chunkCount}:`, JSON.stringify(Object.keys(chunk)));
        
        // Process agent node outputs (LLM responses)
        if (chunk.agent?.messages) {
          debugLog("stream", `Processing ${chunk.agent.messages.length} agent messages`);
          for (const msg of chunk.agent.messages) {
            newMessages.push(msg);
            
            // Log raw message for debugging
            if (DEBUG) {
              debugLog("stream", "Raw message type:", msg.constructor.name);
              debugLog("stream", "Raw message content:", JSON.stringify(msg.content, null, 2));
              if ('tool_calls' in msg) {
                debugLog("stream", "Tool calls:", JSON.stringify((msg as AIMessage).tool_calls, null, 2));
              }
            }
            
            if (msg instanceof AIMessage || msg instanceof AIMessageChunk) {
              debugLog("stream", "AIMessage received, tool_calls:", msg.tool_calls?.length || 0);
              debugLog("stream", "AIMessage content type:", typeof msg.content, Array.isArray(msg.content) ? `(array of ${msg.content.length})` : "");
              
              // Check for thinking/reasoning in the message
              if (debugCallbacks?.onThinking) {
                // Handle extended thinking blocks (Anthropic)
                const content = msg.content;
                if (Array.isArray(content)) {
                  for (const block of content) {
                    if (typeof block === "object" && block !== null) {
                      if ("type" in block && block.type === "thinking" && "thinking" in block) {
                        debugCallbacks.onThinking(block.thinking as string);
                      }
                    }
                  }
                }
              }

              // Check for tool calls
              if (debugCallbacks?.onToolCall && msg.tool_calls) {
                for (const toolCall of msg.tool_calls) {
                  debugCallbacks.onToolCall(toolCall.name, toolCall.args as Record<string, unknown>);
                }
              }

              // Capture final text response
              const textContent = typeof msg.content === "string" 
                ? msg.content 
                : Array.isArray(msg.content)
                  ? msg.content
                      .filter((c): c is { type: "text"; text: string } => 
                        typeof c === "object" && c !== null && "type" in c && c.type === "text"
                      )
                      .map(c => c.text)
                      .join("")
                  : "";
              
              debugLog("stream", "Extracted text content length:", textContent.length);
              
              if (textContent && !msg.tool_calls?.length) {
                finalResponse = textContent;
                debugLog("stream", "Set finalResponse, length:", finalResponse.length);
              }
            }
          }
        }

        // Process tool node outputs (tool results)
        if (chunk.tools?.messages) {
          debugLog("stream", `Processing ${chunk.tools.messages.length} tool messages`);
          for (const msg of chunk.tools.messages) {
            newMessages.push(msg);
            
            if (msg instanceof ToolMessage && debugCallbacks?.onToolResult) {
              const content = typeof msg.content === "string" 
                ? msg.content 
                : JSON.stringify(msg.content);
              debugCallbacks.onToolResult(msg.name || "unknown", content);
            }
          }
        }
      }
      } catch (iterError) {
        debugLog("chat", "ERROR during stream iteration:", iterError);
        console.error("[ITERATION ERROR]", iterError);
        throw iterError;
      }

      debugLog("chat", `Stream complete. Total chunks: ${chunkCount}, finalResponse length: ${finalResponse.length}`);
      
      // Update message history with all new messages
      messageHistory = [...messageHistory, ...newMessages];

      return finalResponse;
    },

    getHistory: () => messageHistory,
  };
}

File: web_client/agent.py
"""LangChain ReAct agent for the RPG game."""

import json
import os
from dataclasses import dataclass, field
from typing import Literal
from langchain_core.messages import HumanMessage, AIMessage, ToolMessage, BaseMessage
from langchain_core.tools import StructuredTool
from langgraph.prebuilt import create_react_agent

SYSTEM_PROMPT = """You are an engaging Dungeon Master running a text-based RPG adventure.

## Communication Modes

Players communicate in TWO modes, indicated by message prefixes:

**[OOC]** (Out-of-Character) - The player is talking to YOU as the game master:
- Answer questions about game mechanics, available tools, rules
- Explain character stats, inventory, abilities
- Discuss strategy, give hints if asked
- Help with session management (creating sessions, checking state)
- Be helpful and informative, break the fourth wall freely
- List available actions or explain how things work

**[IC]** (In-Character) - The player is roleplaying as their character:
- Respond as the world and NPCs would
- Describe scenes, environments, and NPC reactions dramatically
- Stay fully immersed - never break character
- Narrate the results of their actions cinematically
- Roll dice when needed and describe outcomes vividly

## Your Role

- Create immersive descriptions of locations, NPCs, and events
- Guide players through the story using the available tools
- Roll dice for combat, skill checks, and random outcomes
- Keep track of the game state using sessions and character sheets
- Reference world lore for consistent storytelling

## Getting Started with a New Player

1. Create a session with create_session() if one doesn't exist
2. Show them the current map with get_current_map()
3. Describe their surroundings vividly
4. Ask what they'd like to do

When in IC mode, be descriptive, dramatic, and fun!
Use the lookup tools to reference lore and keep the world consistent.
When combat happens, use roll_dice and describe the action cinematically."""


ModelProvider = Literal["openai", "anthropic"]


@dataclass
class DebugEvent:
    """A debug event from the agent."""
    type: Literal["thinking", "tool_call", "tool_result"]
    name: str = ""
    content: str = ""


@dataclass 
class ChatResult:
    """Result from a chat interaction."""
    response: str
    debug_events: list[DebugEvent] = field(default_factory=list)


def create_model(provider: ModelProvider, model_name: str | None = None):
    """Create a chat model based on provider."""
    if provider == "anthropic":
        from langchain_anthropic import ChatAnthropic
        return ChatAnthropic(
            model=model_name or "claude-sonnet-4-20250514",
            temperature=0.7,
        )
    else:
        from langchain_openai import ChatOpenAI
        return ChatOpenAI(
            model=model_name or "gpt-4o",
            temperature=0.7,
        )


def detect_provider() -> tuple[ModelProvider, str | None]:
    """Auto-detect which provider to use based on environment variables."""
    explicit_provider = os.environ.get("LLM_PROVIDER", "").lower()
    explicit_model = os.environ.get("LLM_MODEL")
    
    if explicit_provider in ("anthropic", "openai"):
        return explicit_provider, explicit_model  # type: ignore
    
    has_anthropic = bool(os.environ.get("ANTHROPIC_API_KEY"))
    has_openai = bool(os.environ.get("OPENAI_API_KEY"))
    
    if has_anthropic and not has_openai:
        return "anthropic", explicit_model
    if has_openai and not has_anthropic:
        return "openai", explicit_model
    if has_anthropic and has_openai:
        return "anthropic", explicit_model  # Default to Anthropic
    
    return "openai", explicit_model


class GameAgent:
    """ReAct agent for running the RPG game."""
    
    def __init__(self, tools: list[StructuredTool], provider: ModelProvider | None = None, model_name: str | None = None):
        if provider is None:
            provider, model_name = detect_provider()
        
        self.provider = provider
        self.model_name = model_name
        self.model = create_model(provider, model_name)
        self.tools = tools
        
        self.agent = create_react_agent(
            model=self.model,
            tools=tools,
            prompt=SYSTEM_PROMPT,
        )
        
        self.message_history: list[BaseMessage] = []
    
    async def chat(self, message: str, mode: Literal["ic", "ooc"] = "ic", debug: bool = False) -> ChatResult:
        """Send a message and get a response with optional debug info."""
        # Format message with mode prefix
        prefix = "[OOC]" if mode == "ooc" else "[IC]"
        formatted = f"{prefix} {message}"
        
        self.message_history.append(HumanMessage(content=formatted))
        
        debug_events: list[DebugEvent] = []
        final_response = ""
        
        # Use streaming to capture intermediate steps
        stream = self.agent.astream(
            {"messages": self.message_history},
            stream_mode="updates"
        )
        
        new_messages: list[BaseMessage] = []
        
        async for chunk in stream:
            # Process agent node outputs (LLM responses)
            if "agent" in chunk and chunk["agent"].get("messages"):
                for msg in chunk["agent"]["messages"]:
                    new_messages.append(msg)
                    
                    if isinstance(msg, AIMessage):
                        # Check for thinking blocks (Anthropic extended thinking)
                        if debug:
                            content = msg.content
                            if isinstance(content, list):
                                for block in content:
                                    if isinstance(block, dict) and block.get("type") == "thinking":
                                        debug_events.append(DebugEvent(
                                            type="thinking",
                                            content=block.get("thinking", "")
                                        ))
                        
                        # Check for tool calls
                        if debug and msg.tool_calls:
                            for tool_call in msg.tool_calls:
                                debug_events.append(DebugEvent(
                                    type="tool_call",
                                    name=tool_call["name"],
                                    content=json.dumps(tool_call["args"], indent=2)
                                ))
                        
                        # Capture final text response
                        content = msg.content
                        if isinstance(content, str) and content and not msg.tool_calls:
                            final_response = content
                        elif isinstance(content, list):
                            texts = [c.get("text", "") for c in content 
                                    if isinstance(c, dict) and c.get("type") == "text"]
                            if texts and not msg.tool_calls:
                                final_response = "".join(texts)
            
            # Process tool node outputs (tool results)
            if "tools" in chunk and chunk["tools"].get("messages"):
                for msg in chunk["tools"]["messages"]:
                    new_messages.append(msg)
                    
                    if debug and isinstance(msg, ToolMessage):
                        content = msg.content if isinstance(msg.content, str) else json.dumps(msg.content)
                        debug_events.append(DebugEvent(
                            type="tool_result", 
                            name=msg.name or "unknown",
                            content=content
                        ))
        
        # Update message history
        self.message_history = [*self.message_history, *new_messages]
        
        return ChatResult(response=final_response or "(No response)", debug_events=debug_events)
    
    def clear_history(self):
        """Clear the conversation history."""
        self.message_history = []
